### ISSUE TITLE
Detached Participant audio can still be heard

### ISSUE THREAD

**Issue by ghost** [twilio/twilio-video.js#140]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [ ] I verified that the [Quickstart](https://github.com/twilio/video-quickstart-js) application works in my environment.

**Code to reproduce the issue:**
```js
participantOnOffVolume (instanceId) {
    if (document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.contains('zmdi-volume-off')) this.attachTracks(Array.from(this.videos[instanceId].audioTracks.values()), document.getElementById(`participant-${instanceId}-video`), this.videos[instanceId]);
    else this.detachTracks(Array.from(this.videos[instanceId].audioTracks.values()), this.videos[instanceId]);
  }

detachTracks (tracks, participant) {
    tracks.forEach((track) => {
      track.detach()
        .forEach((detachedElement) => {
          detachedElement.remove();
          this.adjustButtons(participant._instanceId);
        });
    });
  }

```

**Expected behavior:**
Functionality is meant to mute(detach) or unmute(reattach) userA's audio track on userB's browser during a group call.  Upon pressing the mute audio button, userA's audio track should no longer be heard after detached and removed.

**Actual behavior:**
UserA's audio track can be heard by userB even after detached and removed.  HTML has been inspected and found no audio track but still heard audio.

**Software versions:**

 - [x] Browser(s): Chrome 59.0.3071.115
 - [x] Operating System: Win10
 - [x] twilio-video.js: 1.2.0
 - [x] Third-party libraries (e.g., Angular, React, etc.): Angular 1.6.5

**Comment by Samenai** [comment]
Video: https://streamable.com/xlwpk

UserB has camera and mic turned off, userA has camera turned off.  You can hear userA's audio track (music I played on a bluetooth speaker and my typing) even though the track was detached and removed.

**Comment by Samenai** [comment]
Quick workaround:

```js
participantOnOffVolume (instanceId) {
    if (document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.contains('zmdi-volume-off')) {
      document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.remove('zmdi-volume-off');
      document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.add('zmdi-volume-up');
      document.getElementById(`participant-${instanceId}-video`).querySelector('audio').muted = false;
    } else {
      document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.remove('zmdi-volume-up');
      document.getElementById(`participant-${instanceId}-micToggle`).querySelector('#buttonVolume').classList.add('zmdi-volume-off');
      document.getElementById(`participant-${instanceId}-video`).querySelector('audio').muted = true;
    }
  }
```

**Comment by markandrus** [comment]
Hi @Samenai,

Thanks for the code sample and the video. Could you share the source for `attachTracks` and the caller(s) of `participantOnOffVolume`?

> I realize this might not actually be a bug and I'm incorrectly attempting to mute/unmute a user's track by detaching/attaching, but I could not find other options to mute/unmute in the participant object, audio track object, or the twilio-video documentation, other than html muting the audio tag.

Detaching a remote Participant's Track _should_ mute it, so I'm wondering what could be going wrong here. I'll play around with it a bit tomorrow to see if I can repro.

**Comment by Samenai** [comment]
Hi @markandrus,

Here's the source for attachTracks:

```js
attachTracks (tracks, container, participant) {
    tracks.forEach((track) => {
      if (track.kind === 'video' && container.querySelector('video')) container.removeChild(container.querySelector('video'));
      if (track.kind === 'video' && container.querySelector('#disabledAvatar')) container.removeChild(container.querySelector('#disabledAvatar'));
      if (track.kind === 'audio' && container.querySelector('audio')) container.removeChild(container.querySelector('audio'));
      container.appendChild(track.attach());
      if (track.kind === 'video' && participant === this.selectedSpeaker) {
        if (document.getElementById('main-video').querySelector('#disabledAvatar')) document.getElementById('main-video').removeChild(document.getElementById('main-video').querySelector('#disabledAvatar'));
        if (document.getElementById('main-video').querySelector('video')) document.getElementById('main-video').removeChild(document.getElementById('main-video').querySelector('video'));
        document.getElementById('main-video').appendChild(track.attach());
      }
      if (track.kind === 'audio') {
        this.hark(participant);
        if (participant._instanceId !== this.activeRoom.localParticipant._instanceId && document.getElementById('buttonAllVolume').classList.contains('zmdi-volume-off')) {
          document.getElementById('buttonAllVolume').classList.remove('zmdi-volume-off');
          document.getElementById('buttonAllVolume').classList.add('zmdi-volume-up');
        }
      }
      this.adjustButtons(participant._instanceId);
    });
  }
```

And the caller of participantOnOffVolume is an angular ng-click inside a directive, here's a snippet:

```html
<div style="text-align:right; padding:0 !important;" class="smallVidButtonContainer col-md-2 col-sm-2 col-xs-2"
         id="participant-{{instanceId}}-cameraToggle">
      <button ng-if="participant.videoTracks.size > 0 && participant.videoTracks.entries().next().value[1].isEnabled" style="margin:0 !important; background-color:#dfab26;" class="btn btn-xs btn-blue-grey btn-fab zmdi zmdi-videocam zmdi-hc-2x circularButton" ng-click="participantOnOffVideocam(instanceId)" id="buttonVideocam" title="Toggle Camera"></button>
    </div>
```

Looking over attachTracks again, I noticed something.  Typically those 'removeChild's in attachTracks aren't called but maybe that's a possible cause?  Because I'm not 'gracefully' removing any tracks with track.detach() and elem.remove()... I'll look into that next.

**Comment by markandrus** [comment]
How do you set `this.videos[instanceId]` and `this.videos[instanceId].audioTracks`? Is that just the Participant and the Participant's `audioTracks`? I'm wondering if an AudioTrack is "left behind" on the page, maybe because it is missing from that collection, when `detachTracks` is called. Although your video that you shared wouldn't suggest that...

**Comment by Samenai** [comment]
```js
this.videos[this.activeRoom.localParticipant._instanceId] = this.activeRoom.localParticipant;
this.activeRoom.participants.forEach((participant) => {
  this.videos[participant._instanceId] = participant;
  this.createContainer(participant); //participant directive creation
});
this.activeRoom.on('participantConnected', (participant) => {
  this.videos[participant._instanceId] = participant;
  this.createContainer(participant); //participant directive creation
});
this.activeRoom.on('participantDisconnected', (participant) => {
  Reflect.deleteProperty(this.videos, participant._instanceId);
  this.detachParticipantTracks(participant);
});
```

**Comment by markandrus** [comment]
Thanks for your patience, @Samenai. I've reproduced this bug (see [JSFiddle](https://jsfiddle.net/9ybs7var/3/)). It's independent of Angular. We'll work on a fix, but for now, please continue using the workaround you identified.

There are a few issues I see:

1. I could've sworn it used to be the case that removing an &lt;audio&gt; or &lt;video&gt; element from the DOM stopped it from playing. At least that is how it works in Firefox.
2. It's not enough to `null` out the `srcObject` attribute on the HTMLMediaElement after removing it from the DOM; we need to `null` it out before.

Also, one request: can you please use the `sid` attribute of the Participant as a unique identifier instead of `_instanceId`? `_instanceId` is private API, and subject to change in any release (even patch-level). TBH, I forgot it was even a part of our implementation. So I would not rely on it.

Thanks!
Mark

**Comment by markandrus** [comment]
I've opened up a bug against Chrome. This issue is also present in Safari 11 Tech Preview, although I haven't opened a bug against Safari yet.

https://bugs.chromium.org/p/chromium/issues/detail?id=749928

**Comment by Samenai** [comment]
Wow crazy! I'll use the workaround for now and definitely interested to see how this turns out!

No worries on the _instanceId => sid change.  I noticed the use of sid in some quickstart code and was going to make the change soon.

Best of luck with resolving!

**Comment by markandrus** [comment]
I'm reopening this, since I'm not seeing our `srcObject` update work anymore, unfortunately :-( We should continue watching https://bugs.chromium.org/p/chromium/issues/detail?id=749928 for a proper fix from Chrome.

**Comment by xandervdm** [comment]
It is now 2019 and unfortunately detaching the remote audio track still doesn't work under Chrome (remote audio can still be heard), yet it works under FireFox and Safari. Detaching the remote video track works correctly under all three these browsers.

**Comment by makarandp0** [comment]
@xandervdm, I have started looking at this again, and finding that in chrome plays audio track w/o audio element being added to the DOM. 

This jsfiddle illustrates the issue.
https://jsfiddle.net/makarandp/87nzrcsb/13/ 

It looks like that setting `audioElement.srcObject = null` does stop the track from playing. I am trying to see if we can use this to workaround the issue in our SDK.

**Comment by makarandp0** [comment]
We now have a workaround in place for this issue in [beta15](https://github.com/twilio/twilio-video.js/releases/tag/2.0.0-beta15) release. Please check it out, and let us know if that works for you.
